cmd/root.go

```diff
@@ -0,0 +1,129 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/pengsrc/go-shared/check"
+	uuid "github.com/satori/go.uuid"
+	"github.com/spf13/cobra"
+
+	"github.com/beyondstorage/beyond-ftp/client"
+	"github.com/beyondstorage/beyond-ftp/config"
+	"github.com/beyondstorage/beyond-ftp/constants"
+	"github.com/beyondstorage/beyond-ftp/server"
+	"github.com/beyondstorage/beyond-ftp/utils"
+)
+
+var (
+	versionFlag bool
+	cfgFileFlag string
+
+	clientCount      int
+	clientCountMutex sync.Mutex
+)
+
+// RootCmd represents the base command when called without any subcommands
+var RootCmd = &cobra.Command{
+	Use:   constants.Name,
+	Short: "A FTP server that persists all data to Beyond Storage.",
+	Long:  "A FTP server that persists all data to Beyond Storage.",
+	Run: func(cmd *cobra.Command, args []string) {
+		if versionFlag {
+			fmt.Fprintf(os.Stdout, "BeyondFTP version %s\n", constants.Version)
+			return
+		}
+
+		config.LoadConfigFromFilepath(cfgFileFlag)
+		s := server.NewFTPServer()
+		StartServer(s)
+	},
+}
+
+func StartServer(s server.Server) {
+	s.Start()
+	go signalHandler(s)
+	for {
+		connection, addr, err := s.AcceptClient()
+		if err != nil {
+			utils.Logger.Errorf("Accept error: %v", err)
+			return
+		}
+
+		id := strings.Replace(uuid.NewV4().String(), "-", "", -1)
+		go serveClient(s, id, addr, connection)
+	}
+}
+
+func serveClient(s server.Server, id, addr string, connection utils.Conn) {
+	setting := s.Setting()
+	c := client.NewHandler(
+		id, addr, connection, setting,
+		s.DriverFactory, s.PassiveTransferFactory, s.ActiveTransferFactory,
+	)
+
+	if err := clientArrival(id, addr, setting.MaxConnections); err != nil {
+		c.WriteMessage(500, fmt.Sprintf("Can't accept you - %v", err.Error()))
+		return
+	}
+
+	c.WriteMessage(220, "Welcome to BeyondFTP Server")
+	utils.Logger.Debugf("Accept client on: id: %s, IP: %v", id, addr)
+
+	c.HandleCommands()
+
+	utils.Logger.Debugf("Goodbye: id: %s, IP: %v", id, addr)
+	clientDeparture(id, addr)
+}
+
+func clientArrival(id string, addr string, maxConnections int) error {
+	clientCountMutex.Lock()
+	defer clientCountMutex.Unlock()
+
+	if clientCount+1 > maxConnections {
```

Comment:
There is no feature request for `max_connection` support, how about remove them?